packages/cactus-plugin-ledger-connector-polkadot/src/main/typescript/plugin-ledger-connector-polkadot.ts

```diff
@@ -0,0 +1,901 @@
+        "The operation has failed because the API is not connected to Substrate Node",
+      );
+    }
+    const contractAbi = new Abi(
+      req.metadata,
+      this.api.registry.getChainProperties(),
+    );
+    const methodName = stringCamelCase(req.methodName);
+    const isSafeToCall = await this.isSafeToCallContractMethod(
+      contractAbi,
+      methodName,
+    );
+    if (!isSafeToCall) {
+      throw new RuntimeError(
+        `Invalid method name provided in request. ${req.methodName} does not exist on the contract abi.messages object's "method" property.`,
+      );
+    }
+    const contract = new ContractPromise(
+      this.api,
+      req.metadata,
+      req.contractAddress,
+    );
+    const gasLimit: WeightV2 = this.api.registry.createType("WeightV2", {
+      refTime: req.gasLimit.refTime,
+      proofSize: req.gasLimit.proofSize,
+    });
+    if (req.invocationType === PolkadotContractInvocationType.Query) {
+      let success = false;
+      const params = req.params ?? []
+      const query = contract.query[methodName](
+              req.accountAddress,
+              {
+                gasLimit,
+                storageDepositLimit: req.storageDepositLimit,
+                value: req.balance,
+              },
+              ...params,
+            )
+      const callOutput = await query;
+      success = true;
+      return { success, callOutput };
+    } else if (req.invocationType === PolkadotContractInvocationType.Send) {
+      if (isWeb3SigningCredentialNone(req.web3SigningCredential)) {
+        throw new Error(`${fnTag} Cannot invoke contract with pre-signed TX`);
+      }
+      let mnemonic: string;
+      if (isWeb3SigningCredentialMnemonicString(req.web3SigningCredential)) {
+        const web3SigningCredential =
+          req.web3SigningCredential as Web3SigningCredentialMnemonicString;
+        mnemonic = web3SigningCredential.mnemonic;
+      } else if (isWeb3SigningCredentialCactusRef(req.web3SigningCredential)) {
+        const web3SigningCredential =
+          req.web3SigningCredential as Web3SigningCredentialCactusKeychainRef;
+        const { keychainEntryKey, keychainId } = web3SigningCredential;
+        // locate the keychain plugin that has access to the keychain backend
+        // denoted by the keychainID from the request.
+        const keychainPlugin =
+          this.pluginRegistry.findOneByKeychainId(keychainId);
+
+        Checks.truthy(
+          keychainPlugin,
+          `${fnTag} keychain for ID:"${keychainId}"`,
+        );
+
+        // Now use the found keychain plugin to actually perform the lookup of
+        // the private key that we need to run the transaction.
+        mnemonic = await keychainPlugin?.get(keychainEntryKey);
+      } else {
+        throw new Error(
+          `${fnTag} Unrecognized Web3SigningCredentialType: ` +
+            `Supported ones are: ` +
+            `${Object.values(Web3SigningCredentialType).join(";")}`,
+        );
+      }
```

Comment:
Common code like this can be extracted to a separate (private) method (both here and in `deployContract`)

Reply:
done